citest/gcp_testing/api_investigator.py

```diff
@@ -0,0 +1,376 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Collects resources managed by Google APIs."""
+
+import re
+
+from citest.gcp_testing.gcp_agent import GcpAgent
+from citest.base import ExecutionContext
+
+
+class ApiResourceFilter(object):
+  """Determines which resources in an API are wanted."""
+  # pylint: disable=too-few-public-methods
+
+  def __init__(self, include_regexs, exclude_regexs):
+    """Constructor.
+
+    Args:
+      include_regexs: [list of re] Regexes to include.
+      exclude_regexes: [list of re] Regexes to subtract from include.
+    """
+    self.__include = include_regexs
+    self.__exclude = exclude_regexs
+
+  def __repr__(self):
+    return 'ApiResourceFilter\n  %r\n  %r' % (self.__include, self.__exclude)
+
+  def wanted(self, resource_name):
+    """Determine if resource_name is among those included but not excluded."""
+    found = False
+    for exp in self.__include:
+      if exp.match(resource_name):
+        found = True
+        break
+
+    if found:
+      for exp in self.__exclude:
+        if exp.match(resource_name):
+          found = False
+          break
+
+    return found
+
+
+class ApiInvestigator(object):
+  """Investigate APIs for what resources they manage and how to inspect them."""
+
+  __api_version_map = None
+
+  @staticmethod
+  def collect_apis():
+    """Collect names and current versions of all the Google APIs."""
+    if ApiInvestigator.__api_version_map:
+      return ApiInvestigator.__api_version_map
+
+    context = ExecutionContext()
+    agent = GcpAgent.make_agent(api='discovery', version='v1')
+    apis = agent.list_resource(context, 'apis')
+
+    catalog = {}
+    for entry in apis:
+      preferred = entry.get('preferred', True)
+      if not preferred:
+        continue
+      catalog[entry['name']] = entry['version']
+
+    ApiInvestigator.__api_version_map = catalog
+    return catalog
+
+  @staticmethod
+  def get_api_version(api):
+    """Return current version of requested api name."""
+    return ApiInvestigator.collect_apis()[api]
```

Comment:
Can we collect the APIs from the discovery doc once instead of for each API call?

Reply:
collect_apis() is on demand, then caches for future calls. So it already is once as you suggest.